pkg/submitter/submitter.go

```diff
@@ -0,0 +1,500 @@
+// Copyright (c) The Tellor Authors.
+// Licensed under the MIT License.
+
+package submitter
+
+import (
+	"context"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"math/big"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	tellorCommon "github.com/tellor-io/telliot/pkg/common"
+	"github.com/tellor-io/telliot/pkg/config"
+	"github.com/tellor-io/telliot/pkg/contracts"
+	"github.com/tellor-io/telliot/pkg/db"
+	"github.com/tellor-io/telliot/pkg/mining"
+	"github.com/tellor-io/telliot/pkg/rpc"
+	"github.com/tellor-io/telliot/pkg/tracker"
+)
+
+const ComponentName = "submitter"
+
+/**
+* The submitter has one purpose: to either submit the solution on-chain
+* or to reject it if the miner has already submitted a solution for the challenge
+* or the the solution's challenge does not match current challenge
+ */
+
+type Submitter struct {
+	ctx              context.Context
+	close            context.CancelFunc
+	logger           log.Logger
+	cfg              *config.Config
+	proxy            db.DataServerProxy
+	account          *rpc.Account
+	client           contracts.ETHClient
+	contractInstance *contracts.ITellor
+	currentChallenge *mining.MiningChallenge
+	currentNonce     string
+	currentValues    [5]*big.Int
+	submitter        tellorCommon.TransactionSubmitter
+	resultCh         chan *mining.Result
+	submitCount      prometheus.Counter
+	submitFailCount  prometheus.Counter
+	submitProfit     *prometheus.GaugeVec
+	submitCost       *prometheus.GaugeVec
+	submitReward     *prometheus.GaugeVec
+	lastSubmitCncl   context.CancelFunc
+}
+
+func CreateSubmitter(ctx context.Context, cfg *config.Config, logger log.Logger, client contracts.ETHClient, contractInstance *contracts.ITellor, account *rpc.Account, proxy db.DataServerProxy) (*Submitter, chan *mining.Result) {
+	ctx, close := context.WithCancel(ctx)
+	submitter := &Submitter{
+		ctx:              ctx,
+		close:            close,
+		client:           client,
+		proxy:            proxy,
+		cfg:              cfg,
+		resultCh:         make(chan *mining.Result),
+		account:          account,
+		logger:           logger,
```

Comment:
need to add the component name and also the pub key to the logs setup. Also add the pub key in the logs for the mining